storage/store.go

```diff
@@ -1198,97 +1198,100 @@ func (s *Store) Descriptor() (*proto.StoreDescriptor, error) {
-	ctx = s.Context(ctx)
-	// If the request has a zero timestamp, initialize to this node's clock.
-	header := args.Header()
-	if err := verifyKeys(header.Key, header.EndKey, proto.IsRange(call.Args)); err != nil {
-		reply.Header().SetGoError(err)
-		return err
+	argsHeader, replyHeader := args.Header(), reply.Header()
+
+	var err error
+	defer func() {
+		if err != nil {
+			replyHeader.SetGoError(err)
+		}
+	}()
+
+	if err = verifyKeys(argsHeader.Key, argsHeader.EndKey, proto.IsRange(call.Args)); err != nil {
+		return
 	}
-	if !header.Timestamp.Equal(proto.ZeroTimestamp) {
+
+	if !argsHeader.Timestamp.Equal(proto.ZeroTimestamp) {
 		if s.Clock().MaxOffset() > 0 {
 			// Once a command is submitted to raft, all replicas' logical
 			// clocks will be ratcheted forward to match. If the command
 			// appears to come from a node with a bad clock, reject it now
 			// before we reach that point.
-			offset := time.Duration(header.Timestamp.WallTime - s.Clock().PhysicalNow())
+			offset := time.Duration(argsHeader.Timestamp.WallTime - s.Clock().PhysicalNow())
 			if offset > s.Clock().MaxOffset() {
-				err := util.Errorf("Rejecting command with timestamp in the future: %d (%s ahead)",
-					header.Timestamp.WallTime, offset)
-				reply.Header().SetGoError(err)
-				return err
+				err = util.Errorf("Rejecting command with timestamp in the future: %d (%s ahead)",
+					argsHeader.Timestamp.WallTime, offset)
+				return
 			}
 		}
 		// Update our clock with the incoming request timestamp. This
 		// advances the local node's clock to a high water mark from
 		// amongst all nodes with which it has interacted.
-		s.ctx.Clock.Update(header.Timestamp)
+		s.ctx.Clock.Update(argsHeader.Timestamp)
 	}
 
+	ctx = s.Context(ctx)
+
 	// Backoff and retry loop for handling errors.
 	retryOpts := *s.ctx.RangeRetryOptions
 	retryOpts.Tag = fmt.Sprintf("store: %s", args.Method())
-	err := retry.WithBackoff(retryOpts, func() (retry.Status, error) {
+	err = retry.WithBackoff(retryOpts, func() (retry.Status, error) {
 		// Add the command to the range for execution; exit retry loop on success.
 		reply.Reset()
 
 		// Get range and add command to the range for execution.
-		rng, err := s.GetRange(header.RaftID)
+		rng, err := s.GetRange(argsHeader.RaftID)
 		if err != nil {
-			reply.Header().SetGoError(err)
-			return retry.Break, err
-		}
-
-		if err = rng.AddCmd(ctx, proto.Call{Args: args, Reply: reply}, true); err == nil {
 			return retry.Break, err
 		}
 
-		// Maybe resolve a potential write intent error. We do this here
-		// because this is the code path with the requesting client
-		// waiting. We don't want every replica to attempt to resolve the
-		// intent independently, so we can't do it in Range.executeCmd.
-		if wiErr, ok := err.(*proto.WriteIntentError); ok {
-			pushType := proto.PUSH_TIMESTAMP
-			if proto.IsWrite(args) {
-				pushType = proto.ABORT_TXN
+		if err = rng.AddCmd(ctx, proto.Call{Args: args, Reply: reply}, true); err != nil {
```

Comment:
`rng.AddCmd()` has the same issue as this method. Does `rng.AddCmd()` ever return an error that is different from `reply.GoError`? If so, we used to leave the `reply.GoError` alone but this change will overwrite it with the returned error. I'm not sure if that has anything to do with the test failure but I'm having a hard time spotting other behavioral changes.